main.cpp

```diff
@@ -14,9 +14,130 @@
 
 #include <iostream>
 
-using namespace std;
+#include "libflash/fileio.hpp"
+#include "libflash/optimized_writer.hpp"
 
-int main() {
-	cout << "Mender-flash is live!" << endl;
-	return 0;
+#include <getopt.h>
+
+void PrintHelp() {
+	std::cout
+		<< "Usage: mender-flash [-h|--help] [-s|--input-size <INPUT_SIZE>] -i|--input <INPUT_PATH> -o|--output <OUTPUT_PATH>"
+		<< std::endl;
+}
+
+int main(int argc, char *argv[]) {
+	int volumeSize = 0;
+	std::string inputPath;
+	std::string outputPath;
+	const size_t blockSize = 1024 * 1024; // 1MB block size
+
+	while (1) {
+		static struct option long_options[] = {
+			{"help", no_argument, 0, 'h'},
+			{"input-size", required_argument, 0, 's'},
+			{"input", required_argument, 0, 'i'},
+			{"output", required_argument, 0, 'o'},
+			{0, 0, 0, 0}};
+
+		int option_index = 0;
+		int c = getopt_long(argc, argv, "hs:i:o:", long_options, &option_index);
+		if (c == -1) {
+			break;
+		}
+		switch (c) {
+		case 'h':
+			PrintHelp();
+			return 0;
+
+		case 's':
+			volumeSize = atoi(optarg);
+			break;
+
+		case 'i':
+			inputPath = optarg;
+			break;
+
+		case 'o':
+			outputPath = optarg;
+			break;
+
+		case '?':
+			break;
+
+		default:
+			PrintHelp();
+			exit(EXIT_FAILURE);
+		}
+	}
+
+	if (inputPath.empty() || outputPath.empty()) {
+		std::cerr << "Wrong input parameters!" << std::endl;
+		PrintHelp();
+		exit(EXIT_FAILURE);
+	}
+
+	mender::io::File srcFile;
+	mender::io::File dstFile;
+
+	std::shared_ptr<mender::io::FileReader> reader;
+	if (inputPath == "stdin") {
+		srcFile = mender::io::GetInputStream();
+		reader = std::make_shared<mender::io::InputStreamReader>();
+	} else {
+		auto src = mender::io::Open(inputPath);
+		if (!src) {
+			std::cerr << "Failed to open source: " << inputPath << " (" << src.error().message
+					  << ")";
+			exit(EXIT_FAILURE);
+		}
+		srcFile = src.value();
+		reader = std::make_shared<mender::io::FileReader>(srcFile);
+	}
+
+	bool isUBI = false;
+	auto isUBIRes = mender::io::IsUBIDevice(outputPath);
+	if (isUBIRes.has_value()) {
+		isUBI = isUBIRes.value();
+	}
+
+	auto dst = mender::io::Open(
+		outputPath,
+		!isUBI, // read: only for non-UBI volumes
```

Comment:
I have to admit that I don't remember the details of how UBI works. Is it not legal to read from the same device you're writing to?

Reply:
Comment from the 'go' version of mender-flash:
> 		// No optimized writes possible on UBI (Mirza)
> 		// All the bytes have to be written
because of that we don't need to open if for reading

Reply:
Ah, I see! No reason to repeat the research! 🙂